**复习：**在前面我们已经学习了Pandas基础，第二章我们开始进入数据分析的业务部分，在第二章第一节的内容中，我们学习了**数据的清洗**，这一部分十分重要，只有数据变得相对干净，我们之后对数据的分析才可以更有力。而这一节，我们要做的是数据重构，数据重构依旧属于数据理解（准备）的范围。

In [15]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

#### 开始之前，导入numpy、pandas包和数据

In [1]:
# 导入基本库
import numpy as np
import pandas as pd

In [2]:
# 载入data文件中的:train-left-up.csv
text = pd.read_csv("data/train-left-up.csv")
text.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  
0                            Braund, Mr. Owen Harris  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                             Heikkinen, Miss. Laina  
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  
4                           Allen, Mr. William Henry

# 2 第二章：数据重构


### 2.4 数据的合并

#### 2.4.1 任务一：将data文件夹里面的所有数据都载入，观察数据的之间的关系

In [16]:
#写入代码
df = pd.read_csv('train.csv')
df.head()
df.shape

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

(891, 12)

In [5]:
text_left_up = pd.read_csv("data/train-left-up.csv")
text_left_down = pd.read_csv("data/train-left-down.csv")
text_right_up = pd.read_csv("data/train-right-up.csv")
text_right_down = pd.read_csv("data/train-right-down.csv")

In [17]:
#写入代码
text_left_up.head()
text_left_up.shape

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  
0                            Braund, Mr. Owen Harris  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                             Heikkinen, Miss. Laina  
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  
4                           Allen, Mr. William Henry

(439, 4)

In [20]:
text_left_down.head()
text_left_down.shape

PassengerId  Survived  Pclass                                         Name
0          440         0       2       Kvillner, Mr. Johan Henrik Johannesson
1          441         1       2  Hart, Mrs. Benjamin (Esther Ada Bloomfield)
2          442         0       3                              Hampe, Mr. Leon
3          443         0       3                    Petterson, Mr. Johan Emil
4          444         1       2                    Reynaldo, Ms. Encarnacion

(452, 4)

In [21]:
text_right_up.head()
text_right_up.shape

Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S
1  female  38.0      1      0          PC 17599  71.2833   C85        C
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S
3  female  35.0      1      0            113803  53.1000  C123        S
4    male  35.0      0      0            373450   8.0500   NaN        S

(439, 8)

In [22]:
text_right_down.head()
text_right_down.shape

Sex   Age  SibSp  Parch        Ticket    Fare Cabin Embarked
0    male  31.0      0      0    C.A. 18723  10.500   NaN        S
1  female  45.0      1      1  F.C.C. 13529  26.250   NaN        S
2    male  20.0      0      0        345769   9.500   NaN        S
3    male  25.0      1      0        347076   7.775   NaN        S
4  female  28.0      0      0        230434  13.000   NaN        S

(452, 8)

【提示】结合之前我们加载的train.csv数据，大致预测一下上面的数据是什么

【回答】通过df.shape()函数，我们可以看出891行12列的dataframe被从第439行以下第4列以右切分成四个小的dataframe，文件的命名说明了它们在原dataframe的位置。

#### 2.4.2：任务二：使用concat方法：将数据train-left-up.csv和train-right-up.csv横向合并为一张表，并保存这张表为result_up

In [30]:
#写入代码
list_up = [text_left_up,text_right_up] # 两个dataframe放到一张list里面
result_up = pd.concat(list_up,axis=1) # concat的对象要是list，把list里面的元素合并，axis=1表示在垂直方向进行连接
result_up.head()
type(result_up)
result_up.shape

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

pandas.core.frame.DataFrame

(439, 12)

【分析】可以看出将两个原来df上部dataframe横向合并成了一个新的dataframe

In [32]:
# 保存表为result_up
result_up.to_csv('result_up.csv') 

#### 2.4.3 任务三：使用concat方法：将train-left-down和train-right-down横向合并为一张表，并保存这张表为result_down。然后将上边的result_up和result_down纵向合并为result。

In [33]:
#将train-left-down和train-right-down横向合并为一张表，并保存这张表为result_down
list_down=[text_left_down,text_right_down]
result_down = pd.concat(list_down,axis=1)
result_down.head()
result_down.to_csv('result_down.csv') 

PassengerId  Survived  Pclass                                         Name  \
0          440         0       2       Kvillner, Mr. Johan Henrik Johannesson   
1          441         1       2  Hart, Mrs. Benjamin (Esther Ada Bloomfield)   
2          442         0       3                              Hampe, Mr. Leon   
3          443         0       3                    Petterson, Mr. Johan Emil   
4          444         1       2                    Reynaldo, Ms. Encarnacion   

      Sex   Age  SibSp  Parch        Ticket    Fare Cabin Embarked  
0    male  31.0      0      0    C.A. 18723  10.500   NaN        S  
1  female  45.0      1      1  F.C.C. 13529  26.250   NaN        S  
2    male  20.0      0      0        345769   9.500   NaN        S  
3    male  25.0      1      0        347076   7.775   NaN        S  
4  female  28.0      0      0        230434  13.000   NaN        S

In [36]:
# 将上边的result_up和result_down纵向合并为result
result = pd.concat([result_up,result_down]) #默认axis=0，即水平方向拼接
result.head()
type(result)
result.shape

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

pandas.core.frame.DataFrame

(891, 12)

In [ ]:
【分析】可以看出，上部dataframe和下部dataframe纵向合并成功

#### 2.4.4 任务四：使用DataFrame自带的方法join方法和append：完成任务二和任务三的任务

In [38]:
resul_up = text_left_up.join(text_right_up)
result_down = text_left_down.join(text_right_down)
result = result_up.append(result_down)
result.head()
type(result)
result.shape

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

pandas.core.frame.DataFrame

(891, 12)

【分析】join可以做横向操作，append可以做纵向操作，具体参数和用法见下面参考资料。

#### 2.4.5 任务五：使用Panads的merge方法和DataFrame的append方法：完成任务二和任务三的任务

In [56]:
result_up = pd.merge(text_left_up,text_right_up,left_index=True,right_index=True) # 使用左侧与右侧dataframe的行索引作为连接键
result_down = pd.merge(text_left_down,text_right_down,left_index=True,right_index=True)
result = resul_up.append(result_down)
result.head()
type(result)
result.shape

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

pandas.core.frame.DataFrame

(891, 12)

【思考】对比merge、join以及concat的方法的不同以及相同。思考一下在任务四和任务五的情况下，为什么都要求使用DataFrame的append方法，如果只要求使用merge或者join可不可以完成任务四和任务五呢？

【回答】
- merge、join以及cocat都能进行横向操作。
- concat通过设置axis=1还可以进行纵向操作。merge和concat是pandas的方法，join是DataFrame的方法。
- 任务四和任务五因为分别用merge和join只进行了横向操作，没进行纵向操作，而append可以进行纵向操作。如果只使用merge和join应该也能完成任务四和任务五，还在尝试。

【参考】[DataFrame 数据合并，连接（merge,join,concat)](https://blog.csdn.net/zutsoft/article/details/51498026)  
&emsp;&emsp;&emsp;&emsp; [python 把几个DataFrame合并成一个DataFrame——merge,append,join,concat](https://blog.csdn.net/milton2017/article/details/54406482)

#### 2.4.6 任务六：完成的数据保存为result.csv

In [42]:
result.to_csv('result.csv')

### 2.5 换一种角度看数据

#### 2.5.1 任务一：将我们的数据变为Series类型的数据

---
这个stack函数是干什么的？

In [44]:
# 将完整的数据加载出来
text = pd.read_csv('result.csv')
text.head()

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   
3           3            4         1       1   
4           4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [48]:
unit_result=text.stack().head(20)
unit_result
unit_result.shape
type(unit_result)

0  Unnamed: 0                                                     0
   PassengerId                                                    1
   Survived                                                       0
   Pclass                                                         3
   Name                                     Braund, Mr. Owen Harris
   Sex                                                         male
   Age                                                           22
   SibSp                                                          1
   Parch                                                          0
   Ticket                                                 A/5 21171
   Fare                                                        7.25
   Embarked                                                       S
1  Unnamed: 0                                                     1
   PassengerId                                                    2
   Survived                                     

(20,)

pandas.core.series.Series

【分析】可以看出，stack: 将数据从DataFrame变成层次化的Series，即将其列索引变成行索引。

【参考】[https://www.lizenghai.com/archives/3403.html](https://www.lizenghai.com/archives/3403.html)

In [51]:
#将代码保存为unit_result,csv
unit_result.to_csv('unit_result.csv')

E:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.
  


In [52]:
test = pd.read_csv('unit_result.csv')

In [55]:
test.head()
type(test)

0   Unnamed: 0                      0.1
0  0  PassengerId                        1
1  0     Survived                        0
2  0       Pclass                        3
3  0         Name  Braund, Mr. Owen Harris
4  0          Sex                     male

pandas.core.frame.DataFrame